File: src/main/java/chapter5/example/CallbackHell.java
package chapter5.example;

import common.CommonUtils;
import common.Log;
import okhttp3.*;

import java.io.IOException;


public class CallbackHell {

    private static final String FIRST_URL = "https://api.github.com/zen";
    private static final String SECOND_URL = CommonUtils.GITHUB_ROOT + "/samples/callback_hell";

    private final OkHttpClient client = new OkHttpClient();

    private Callback onSuccess = new Callback() {
        @Override
        public void onFailure(Call call, IOException e) {
            e.printStackTrace();
        }

        @Override
        public void onResponse(Call call, Response response) throws IOException {
            Log.it(response.body().string());
        }
    };


    public void run() {
        CommonUtils.exampleStart();
        Request request = new Request.Builder()
                .url(FIRST_URL)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                e.printStackTrace();
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                Log.it(response.body().string());

                Request request = new Request.Builder()
                        .url(SECOND_URL)
                        .build();
                client.newCall(request).enqueue(onSuccess);
            }
        });
    }

    public static void main(String[] args) {
        CallbackHell demo = new CallbackHell();
        demo.run();
    }

}

File: src/main/kotlin/chapter5/example/CallbackHell_.kt
package chapter5.example

import common.CommonUtils_
import common.Log_
import okhttp3.*
import java.io.IOException


class CallbackHell_ {

    private val FIRST_URL = "https://api.github.com/zen"
    private val SECOND_URL = CommonUtils_.GITHUB_ROOT + "/samples/callback_hell"

    private val client = OkHttpClient()

    private val onSuccess = object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            e.printStackTrace()
        }

        @Throws(IOException::class)
        override fun onResponse(call: Call, response: Response) {
            Log_.it(response.body().string())
        }
    }


    fun run() {
        CommonUtils_.exampleStart()
        val request = Request.Builder()
                .url(FIRST_URL)
                .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                e.printStackTrace()
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                Log_.it(response.body().string())

                val request = Request.Builder()
                        .url(SECOND_URL)
                        .build()
                client.newCall(request).enqueue(onSuccess)
            }
        })
    }

}

fun main(args: Array<String>) {
    val demo = CallbackHell_()
    demo.run()
}
